[PHYW-431] [PHYW-432] PCI checks with uBSA

serranozack opened the pull request:
[PHYW-431] Add PCI to uBSA request
[PHYW-432] Error Check PCI

serranozack committed:
otdoa: pass PCI as part of uBSA request

adds pci parameter to ubsa request function

Signed-off-by: Zack Mudd <zmudd@serranosystems.com>

serranozack committed:
otdoa: fix formatting errors

this fixes the compliance check

Signed-off-by: Zack Mudd <zmudd@serranosystems.com>

nriedel commented on include/otdoa_al/otdoa_nordic_at_h1.h:
```diff
@@ -35,20 +38,23 @@
 					    uint32_t *pu32_ecgi, uint32_t *pu32_dlearfcn,
-					    uint16_t *pu16_mcc, uint16_t *pu16_mnc);
+					    uint16_t *pu16_mcc, uint16_t *pu16_mnc,
+					    uint16_t *pu16_pci);
```
Nordic's static code analyzer is not happy with this many arguments to the function.  Let's change this function and `otdoa_nordic_at_get_ecgi_and_dlearfcn()` to pass in a pointer to a structure where the results are written.

nriedel commented on lib/otdoa_al/otdoa_http_api.c:
```diff
@@ -77,29 +77,38 @@ int32_t otdoa_api_ubsa_download(const otdoa_api_ubsa_dl_req_t *p_dl_request,
 			return rc;
 		}
+	} else {
+		/* otherwise, just get the PCI */
```
The PCI needs to be added to the `otdoa_api_ubsa_download` structure.
This API is used by the application to request a download (e.g. from the shell).  If they pass in 0 for ECGI, then get the values from the modem.  Otherwise, just use the values that are passed in.  Don't get the PCI here.

nriedel commented on lib/otdoa_al/otdoa_nordic_at_h1.c:
```diff
@@ -94,6 +95,7 @@ int otdoa_nordic_at_parse_xmonitor_response(const char *const psz_resp, size_t u
 	uint32_t u32_AcT = 0; /* AcT value, 9=>NBIot, 7=>LTE */
 	uint16_t u16_mcc = 0;
 	uint16_t u16_mnc = 0;
+	uint16_t u16_pci = 0;
```
Let's default these to an "unknown" value (e.g. 0xFFFF).  Make sure to define that value with a macro.
Do the same thing with the DLEARFCN.

nriedel commented on lib/otdoa_al/otdoa_nordic_at_h1.c:
```diff
@@ -174,6 +176,16 @@ int otdoa_nordic_at_parse_xmonitor_response(const char *const psz_resp, size_t u
+		{
+			int i_scn_rv = sscanf(token, "%" SCNu16, &u16_pci);
+
+			if (i_scn_rv != 1) {
```
I think there are cases where the PCI may not be present.  These are similar to cases where the DLEARFCN is not present.  So we should just return the "unknown" PCI value in this case.

nriedel commented on lib/otdoa_al/otdoa_shell.c:
```diff
@@ -42,7 +42,7 @@ static int otdoa_shell_info_handler(const struct shell *shell, size_t argc, char
 	shell_print(shell, "Nordic OTDOA Application");
 	/* get the ECGI */
-	rc = otdoa_nordic_at_get_ecgi_and_dlearfcn(&ecgi, NULL, NULL, NULL);
+	rc = otdoa_nordic_at_get_ecgi_and_dlearfcn(&ecgi, NULL, NULL, NULL, NULL);
```
Make sure to handle the case where the DLEARFCN or PCI is not present.  This should not print an error.

nriedel commented on lib/otdoa_al/otdoa_shell.c:
let's add the MNC, MCC, PCI and DLEARFCN here.

nriedel reviewed (changes requested):
OK, good stuff.  A little rework needed, mostly due to Nordic's static code analysis.

nriedel commented on lib/otdoa_al/otdoa_nordic_at_h1.c:
```diff
@@ -82,7 +82,8 @@ char *otdoa_nordic_at_strtok_r(char *s, char delim, char **save_ptr)
 					    uint32_t *pu32_ecgi, uint32_t *pu32_dlearfcn,
-					    uint16_t *pu16_mcc, uint16_t *pu16_mnc)
+					    uint16_t *pu16_mcc, uint16_t *pu16_mnc,
+					    uint16_t *pu16_pci)
```
The Nordic static code analysis want the complexity of this function reduced.  You can find the comment here: https://sonarcloud.io/project/issues?pullRequest=25090&id=nrfconnect_sdk-nrf&open=AZnu_q6cWgIh_fCdIlc_

Let's go ahead and fix that in this PR as well.

serranozack committed:
lib: otdoa: replace xmonitor paramters with struct

Removes ecgi, dlearfcn, mcc, mnc, pci as individual parameters
to functions using the AT%XMONITOR command

Signed-off-by: Zack Mudd <zmudd@serranosystems.com>

serranozack committed:
lib: otdoa: fix compliance issues

compliance requres a body line

Signed-off-by: Zack Mudd <zmudd@serranosystems.com>

serranozack committed:
lib: otdoa: handle missing PCI in xmonitor results

adds special handling for a missing PCI, similar to DLEARFCN

Signed-off-by: Zack Mudd <zmudd@serranosystems.com>

serranozack committed:
lib: otdoa: use separate values for unknown and default pci/earfcn

Displays N/A in phywi info when a PCI or DLEARFCN is missing from the
xmonitor modem response, and uses a default value for a ubsa request

Signed-off-by: Zack Mudd <zmudd@serranosystems.com>

serranozack committed:
lib: otdoa: support optional dlearfcn and pci in ubsa request

if the PCI or DLEARFCN is not available from the modem, do not include
them in the ubsa auth request.

Signed-off-by: Zack Mudd <zmudd@serranosystems.com>

serranozack commented:
Closing as this was cherry-picked into `main-otdoa-al-integration` with 50055ddafa52bdbbb652ad2e073a06471cc96a55
